pep-9999.rst

```diff
@@ -0,0 +1,294 @@
+function containing the comprehension. E.g. given the following function::
+
+  def f(lst):
+      return [locals() for x in lst]
+
+Calling ``f([1])`` in current Python will return::
+
+  [{'.0': <list_iterator object at 0x7f8d37170460>, 'x': 1}]
+
+Where ``.0`` is the internal implementation detail of the synthetic sole
+argument to the comprehension "function."
+
+Under this PEP, it will instead return::
+
+  [{'lst': [1], 'x': 1}]
+
+Now including the outer ``lst`` variable as a local, and eliminating the
+synthetic ``.0``.
+
+no comprehension frame in tracebacks
+------------------------------------
+
+Under this PEP, a comprehension will no longer have its own dedicated frame in a
+stack trace. For example, given this function::
+
+  def f():
+    def g():
+      raise RuntimeError("boom")
+
+    return [g() for x in [1]]
+
+Currently calling ``f()`` results in the following traceback:
+
+.. code-block:: text
+
+   Traceback (most recent call last):
+     File "<stdin>", line 1, in <module>
+     File "<stdin>", line 4, in f
+     File "<stdin>", line 4, in <listcomp>
+     File "<stdin>", line 3, in g
+   RuntimeError: boom
+
+Note the dedicated frame for ``<listcomp>``.
+
+Under this PEP, the traceback looks like this instead:
+
+.. code-block:: text
+
+   Traceback (most recent call last):
+     File "<stdin>", line 1, in <module>
+     File "<stdin>", line 4, in f
+     File "<stdin>", line 3, in g
+   RuntimeError: boom
+
+There is no longer an extra frame for the list comprehension. The frame for the
+``f`` function has the correct line number for the comprehension, however, so
+this simply makes the traceback more compact without losing any useful
+information.
+
+It is theoretically possible that code using warnings with the ``stacklevel``
+argument could observe a behavior change due to the frame stack change. In
+practice, however, this seems unlikely. It would require a warning raised in
+library code that is always called through a comprehension in that same
+library, where the warning is using a ``stacklevel`` of 3+ to bypass the
+comprehension and its containing function and point to a calling frame outside
+the library. In such a scenario it would usually be simpler and more reliable
+to raise the warning closer to the calling code and bypass fewer frames.
+
+
+UnboundLocalError instead of NameError
+--------------------------------------
+
+Although the value of the comprehension iteration variable is saved and restored
+to provide isolation, it still becomes a local variable of the outer function
+under this PEP. Thus, if the comprehension iteration variable is accessed
+outside the comprehension without ever being set outside the comprehension, as
+in this function::
+
+   def f(lst):
+       items = [x for x in lst]
+       return x
+
+Under this PEP, calling ``f()`` will raise an ``UnboundLocalError``, whereas
+currently it raises a ``NameError``.
```

Comment:
It could be noted that ``UnboundLocalError`` is a subclass of ``NameError``, so existing code will continue to work unchanged (unless it explicitly checks the exact exception type, which is usually a bad idea and presumably quite uncommon).